rust/src/cxx_bridge_gobject.rs

```diff
@@ -0,0 +1,53 @@
+//! Wrappers that bridge the world of cxx.rs and GObject-introspect based crate bindings.
+//! See https://github.com/dtolnay/cxx/issues/544
+//! While cxx.rs supports including externally-bound types (like ostree::Repo),
+//! two things make this more complicated.  First, cxx.rs requires implementing
+//! a trait, and due to the orphan rule we can't implement foreign traits on
+//! foreign types.  Second, what we *actually* want on the Rust side isn't
+//! the *_sys type (e.g. ostree_sys::OstreeRepo) but the `ostree::Repo` type.
+//! So for now, we define a `FFIGObjectWrapper` trait that helps with this.
+//! In the future though hopefully cxx.rs improves this situation.
+
+use cxx::{type_id, ExternType};
+use paste::paste;
+
+/// A custom trait used to translate a *_sys C type wrapper
+/// to its GObject version.
+pub trait FFIGObjectWrapper {
+    type Wrapper;
+
+    fn gobj_wrap(&mut self) -> Self::Wrapper;
```

Comment:
Reading through the consumers of this I wonder whether this is missing a `Pin` constraint for its input argument?

Reply:
Hmm.  We indeed could require it for our use case probably just to be safe, but this is wrapping a GObject and the Rust wrapper type increments the refcount.  So it's more like `Clone` on an `Rc`.

Reply:
Also I think this method should be part of gtk-rs...I spent a few minutes looking at using current git for that, fighting 
```
$ cargo build
warning: Patch `gio v0.13.0 (/var/srv/walters/src/github/gtk-rs/gtk-rs/gio)` was not used in the crate graph.
Patch `gio-sys v0.13.0 (/var/srv/walters/src/github/gtk-rs/gtk-rs/gio/sys)` was not used in the crate graph.
Patch `glib v0.13.0 (/var/srv/walters/src/github/gtk-rs/gtk-rs/glib)` was not used in the crate graph.
Patch `glib-sys v0.13.0 (/var/srv/walters/src/github/gtk-rs/gtk-rs/glib/sys)` was not used in the crate graph.
Check that the patched package version and available features are compatible
with the dependency requirements. If the patch has a different version from
what is locked in the Cargo.lock file, run `cargo update` to use the new
version. This may also occur with an optional dependency that is not enabled.
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
```

before figuring out that at least the generated ostree crate is requiring the previous 0.10 version, so any work here is going to block on updating that too.

Reply:
https://github.com/gtk-rs/gtk-rs/pull/246